### PR TITLE
Make sure CheckAndThrow is invoked only within ambient Cancellable context

### PR DESCRIPTION
**Pull request by majocha**
See: https://github.com/dotnet/fsharp/pull/18002#issuecomment-2482900236

Basically, when `Cancelable.CheckAndThrow()` throws the exception, we don't want the exception to leak.
It should be caught inside `Cancellable.run` here:
https://github.com/dotnet/fsharp/blob/e44ae1daecf51d423f0ee187ef97ef6cf6c6dd9a/src/Compiler/Utilities/Cancellable.fs#L44-L52

The idea of `CheckAndThrow` is that it is invoked from code that is unaware of `Cancellable` CE somewhere deeper on the callstack.
This PR checks the static `AsyncLocal<CancellationToken voption>` token. If it is default VNone, we know the invocation is outside of Cancellable. If there is a token present, we know `Cancellable.run` deeper on the stack has set it and we're good.

With this PR, internal error will occur whenever `CheckAndThrow` is called outside of Cancellable CE context.
- [x] Added test.
- [x] Runtime check can be disabled by setting DISABLE_CHECKANDTHROW_ASSERT env variable. I tested manually that it works.

### CODE REVIEW

File: tests/FSharp.Test.Utilities/XunitHelpers.fs [hunk 1]
```diff
@@ -14,6 +15,7 @@ type FSharpXunitFramework(sink: IMessageSink) =
         // right at the start of the test run is here in the constructor.
         // This gets executed once per test assembly.
         MessageSink.sinkWriter |> ignore
+        Cancellable.EnsureCheckAndThrowInvokedWithAmbientCancellable()
```
Comment (majocha) [inline tests/FSharp.Test.Utilities/XunitHelpers.fs:18]: This just installs the check globally for the whole test run.

**Comment by auduchinok** [comment]
Maybe it'd make sense to have this assert not in the tests only? I can imagine some FCS APIs that may trigger analysis and that don't take a cancellation token. If there're no tests for such APIs, it's going to be very difficult to uncover these, as they'd only manifest themself when cancelled at some unfortunate moment. The sooner we catch all these exceptions, the better.

Maybe we could have some global property or other way to control it. With this way we'd enable these asserts in the internal and EAP Rider builds, but could turn it off just in case in the release ones, at least for the time being.

**Comment by majocha** [comment]
TBH my initial thought was to just put a hard `failwith` where the assert is. It would surface as a internal error (see the failing tests) if `CheckAndThrow` is misused anywhere. This would be immediately discoverable.

**Comment by majocha** [comment]
Added DISABLE_CHECKANDTHROW_ASSERT env variable that can be set in case of emergencies.

**Comment by psfinaki** [comment]
@majocha do I get it right that the TODO in the description is already done?

**Comment by majocha** [comment]
> @majocha do I get it right that the TODO in the description is already done?

Yes, forgot to edit this.

**Review by T-Gro — APPROVED** [review]

**Pull request merged**